### DISCUSSION TITLE
Convert ktx2 to png/jpg

### DISCUSSION THREAD

**Discussion by nguyen-lc** [KhronosGroup/KTX-Software#761] [Q&A] [answered]
Do you guys know how to convert ktx2 to png/jpg. Thanks

**Comment by MarkCallow** [comment, marked as answer]
png and jpg files are not equivalent to ktx2 files. Each can only contain a single image while ktx2 files can have many images. Therefore you cannot convert them. You can extract one or more images from a ktx2 file, writing each it to a png file using the `ktx extract` command in the 4.3-alpha1 release.

**Reply by nguyen-lc** [reply]
Can u guild me how to using ktx command. I dont see a guild. Its show me an error 'ktx' is not recognized as an internal or external command, operable program or batch file.

**Reply by MarkCallow** [reply]
Did you download and install [v4.3.0-alpha3](https://github.com/KhronosGroup/KTX-Software/releases/tag/v4.3.0-alpha3) for your OS? You are on Windows, right? Once the software is installed ktx should be found unless during install you unchecked the box to add KTX-Software to your PATH. If you did uncheck it you will need to manually add `C:\Program Files\KTX-Software\bin` to your PATH. After this, and after the install, you may need to restart your command prompt or powershell.

**Reply by nguyen-lc** [reply]
Yes its working but idk how many image in ktx2 file. Do u know how to check it?

**Reply by MarkCallow** [reply]
`ktx help` or `ktx -h` will print an overview of the available commands. `ktx help extract` will give you help for _extract_. `ktx info <filename>` will print information about <filename> including number of layers, faces and levels. 

Note that if the file contains images in one of the floating point formats, you can only extract to .exr files as .png is not capable of holding a floating point HDR image.

**Reply by nguyen-lc** [reply]
Do u know any other way to export image better resolution. it quite blur and low-res

**Reply by MarkCallow** [reply]
Images are exported as is* to png which is lossless. If the output is blurry and small, the input was. What is the format of the texture and what mip level are you exporting?

* Uncompressed images are exported as is. ASTC images are decoded to RGBA and exported. Transcodable images (ETC1S/Basis-LZ or UASTC) are transcoded to RGBA and exported.

**Reply by nguyen-lc** [reply]
can you check for me? i'll send ktx2 file to you

**Reply by MarkCallow** [reply]
`ktx info` will tell you the size of the base level and the number of levels. From that the sizes of other levels are easily calculated. What levels, etc were extracted depends on the command options you used and the input file. Do you have multiple output .png files or just one? If multiple, which file are you looking at?

**Reply by nguyen-lc** [reply]
My ktx2 file using 10 level. I'm just wanna get the first level with one png/jpg image. 
vkFormat: VK_FORMAT_UNDEFINED
typeSize: 1
pixelWidth: 512
pixelHeight: 512
pixelDepth: 0
layerCount: 0
faceCount: 1
levelCount: 10
supercompressionScheme: KTX_SS_BASIS_LZ
dataFormatDescriptor.byteOffset: 0x140
dataFormatDescriptor.byteLength: 44
keyValueData.byteOffset: 0x16c
keyValueData.byteLength: 120
supercompressionGlobalData.byteOffset: 0x1e8
supercompressionGlobalData.byteLength: 10961
This is some info of it

**Reply by MarkCallow** [reply]
If the output file is less than 512x512 one of the following is true:

- You are not looking at the image extracted from level0.
- There is a bug in `ktx extract`.

What was the full `ktx extract` command that you used?

The images in the source ktx are compressed with ETC1S/Basis-LZ. Whatever artifacts are in the source image will be visible in the exported file. To reduce those artifacts you will need to re-encode the original image with different settings.

The tool does not export to jpg as jpg would most likely add further artifacts to the image.

**Reply by nguyen-lc** [reply]
i'm using ktx extract file.ktx2 ./file.jpg

**Reply by MarkCallow** [reply]
> i'm using ktx extract file.ktx2 ./file.jpg

Regardless of the file name, a png file will be output and it will be in a file called `./file.jpg.png`. Is the output file 512x512? Try `ktx extract --level 0 ./file.png` and see if the resulting image is any different.
